kernel/src/hil/display.rs

```diff
@@ -0,0 +1,376 @@
+}
+
+pub trait TextBufferClient {
+    /// The driver calls this function when any command (but a write one)
+    /// finishes executing.
+    fn command_complete(&self, r: Result<(), ErrorCode>);
+
+    /// The driver calls this function when a write command finishes executing.
+    fn write_complete(&self, buffer: &'static mut [u8], len: usize, r: Result<(), ErrorCode>);
+}
+
+/* FrameBuffer */
+
+pub trait FrameBuffer<'a> {
+    /// Returns a tuple (width, height) with the current resolution (in pixels)
+    /// This function is synchronous as the driver should know this value without
+    /// requesting it from the screen.
+    ///
+    /// note that width and height may change due to rotation
+    fn get_mode(&self) -> GraphicsMode;
+
+    /// Returns the minim size (in bytes) that can be written to the framebuffer.
+    ///
+    /// Due to hardware constraints, framebuffer writes have to be rounded up to a tile size.
+    /// This means that the size of the write buffer has to be a multiple of the tile. An
+    /// example use case is a framebuffer that has the minimum write unit a full line on
+    /// the display. This means that clients can only write entier lines to the framebuffer
+    /// as opossed to single pixels.
+    fn get_block_size(&self) -> usize;
+
+    /// Sets the video memory frame.
+    /// This function has to be called before the first call to the write function.
+    /// This will generate a `command_complete()` callback when finished.
+    ///
+    /// Return values:
+    /// - `Ok(())`: The write frame is valid.
+    /// - `INVAL`: The parameters of the write frame are not valid.
+    /// - `BUSY`: Unable to set the write frame on the device.
+    fn set_write_frame(
+        &self,
+        x: usize,
+        y: usize,
+        width: usize,
+        height: usize,
+    ) -> Result<(), ErrorCode>;
+
+    /// Sends a write command to write data in the selected video memory frame.
+    /// When finished, the driver will call the `write_complete()` callback.
+    ///
+    /// Return values:
+    /// - `Ok(())`: Write is valid and will be sent to the screen.
+    /// - `INVAL`: Write is invalid or length is wrong.
+    /// - `BUSY`: Another write is in progress.
+    fn write(
+        &self,
+        buffer: &'static mut [u8],
+        len: usize,
+        reset_position: bool,
+    ) -> Result<(), ErrorCode>;
+
+    /// Flush the framebuffer changes to the hardware device.
+    ///
+    /// Some framebuffers keep in a temporary memory the changes and require a flush command
+    /// to send the changes to the hardware.
+    ///
+    /// Return values:
+    /// - `Ok(())`: Flush is in progress and the client will receive
+    ///    a call to `command_complete`.
+    /// - `ENOSUPPORT`: Flush has been done synchronous or there is no
+    ///    no need to flush the framebuffer.
+    /// - `BUSY`: Another write or flush is in progress.
+    fn flush(&self) -> Result<(), ErrorCode>;
```

Comment:
With https://github.com/tock/tock/pull/3067 I will no longer have any need for flushing.